packages/@lwc/engine-core/src/framework/rendering.ts

```diff
@@ -89,7 +89,15 @@ function patch(n1: VNode, n2: VNode, parent: ParentNode, renderer: RendererAPI)
     }
 
     if (process.env.NODE_ENV !== 'production') {
-        if (!isSameVnode(n1, n2)) {
+        if (
+            !isSameVnode(n1, n2) &&
+            // Currently the only scenario when patch does not receive the same vnodes are for
+            // dynamic components. This is why the final check here ensures n2.type is a custom element.
+            // When a dynamic component's constructor changes, the value of its tag name (sel) will be
+            // different. The engine will unmount the previous element and mount the new one using the
+            // new constructor in patchCustomElement.
+            n2.type !== VNodeType.CustomElement
```

Comment:
I tried to make this comment as clear as possible but what I was going for was, we have a way to handle when the vnodes are not the same (same `sel` and `key`) for custom elements only.  We don't for the other types so we throw an error on those ones.

Reply:
In this case, don't we also expect `n1.type` to be `VNodeType.CustomElement`?

Reply:
Ah you're right, I was thinking it would be enough to only check that `n2.type` is a custom element because it's what's changing and we'd be replacing it in `patchCustomElement`.  But if `n1.type` is not a custom element, `patchCustomElement` will break as well.

I've updated the validation logic in the latest commit.

Reply:
I've updated the logic so this should no longer be needed, see my previous [comment](https://github.com/salesforce/lwc/pull/3338#discussion_r1120347309).